src/vma/dev/qp_mgr_mp.cpp

```diff
@@ -0,0 +1,258 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <dev/qp_mgr_mp.h>
+#include "dev/cq_mgr_mp.h"
+
+#undef  MODULE_NAME
+#define MODULE_NAME 		"qpmp"
+
+#define qp_logpanic 		__log_info_panic
+#define qp_logerr		__log_info_err
+#define qp_logwarn		__log_info_warn
+#define qp_loginfo		__log_info_info
+#define qp_logdbg		__log_info_dbg
+#define qp_logfunc		__log_info_func
+#define qp_logfuncall		__log_info_funcall
+
+
+#ifdef HAVE_MP_RQ
+
+
+cq_mgr* qp_mgr_mp::init_rx_cq_mgr(struct ibv_comp_channel* p_rx_comp_event_channel)
+{
+	uint32_t cq_size = align32pow2(((1 << m_p_ring->get_strides_num()) *
+				       m_p_ring->get_wq_count()));
+	return new cq_mgr_mp(m_p_ring, m_p_ib_ctx_handler, cq_size,
+			     p_rx_comp_event_channel, true,
+			     m_p_ring->get_stride_size());
+}
+
+int qp_mgr_mp::prepare_ibv_qp(vma_ibv_qp_init_attr& qp_init_attr)
+{
+	vma_ibv_device_attr& r_ibv_dev_attr =
+			m_p_ib_ctx_handler->get_ibv_device_attr();
+	NOT_IN_USE(qp_init_attr);
+	// create RX resources
+	// create WQ
+	struct ibv_exp_wq_init_attr wq_init_attr;
+	memset(&wq_init_attr, 0, sizeof(wq_init_attr));
+
+	wq_init_attr.wq_type = IBV_EXP_WQT_RQ;
+	wq_init_attr.max_recv_wr = m_p_ring->get_wq_count();
+	wq_init_attr.max_recv_sge = 1;
+	wq_init_attr.pd = m_p_ib_ctx_handler->get_ibv_pd();
+	wq_init_attr.cq = m_p_cq_mgr_rx->get_ibv_cq_hndl();
+	if (r_ibv_dev_attr.wq_vlan_offloads_cap &
+	    IBV_EXP_RECEIVE_WQ_CVLAN_STRIP) {
+		wq_init_attr.comp_mask |= IBV_EXP_CREATE_WQ_VLAN_OFFLOADS;
+		wq_init_attr.vlan_offloads |= IBV_EXP_RECEIVE_WQ_CVLAN_STRIP;
+	}
+	wq_init_attr.comp_mask |= IBV_EXP_CREATE_WQ_RES_DOMAIN;
+	wq_init_attr.res_domain = m_p_ring->get_res_domain();
+
+	wq_init_attr.comp_mask |= IBV_EXP_CREATE_WQ_MP_RQ;
+	wq_init_attr.mp_rq.use_shift = IBV_EXP_MP_RQ_NO_SHIFT;
+	wq_init_attr.mp_rq.single_wqe_log_num_of_strides =
+					m_p_ring->get_strides_num();
+	wq_init_attr.mp_rq.single_stride_log_num_of_bytes =
+					m_p_ring->get_stride_size();
+
+	m_p_wq = ibv_exp_create_wq(m_p_ib_ctx_handler->get_ibv_context(),
+			&wq_init_attr);
+	if (!m_p_wq) {
+		qp_logerr("ibv_exp_create_wq failed (errno=%d %m)", errno);
+		return -1;
+	}
+
+	// change WQ to ready state
+	struct ibv_exp_wq_attr wq_attr;
+
+	memset(&wq_attr, 0, sizeof(wq_attr));
+	wq_attr.attr_mask = IBV_EXP_WQ_ATTR_STATE;
+	wq_attr.wq_state = IBV_EXP_WQS_RDY;
+
```

Comment:
In all subsequent failures it is necessary to perform cleanup before returning -1.